progresso/__init__.py

```diff
@@ -0,0 +1,216 @@
+        """Estimated time of arrival."""
+        remaining_time = timedelta(seconds=self.tta)
+        return ETA.from_datetime(datetime.now() + remaining_time)
+
+    @property
+    def speed(self):
+        """Number of iterations per second."""
+        return Float(1.0 / self.avg if self.avg else 0)
+
+    def render(self):
+        if self.done < self._last_render + self.throttle <= self.total:
+            return
+        self._last_render = self.done
+        if self.start is None:
+            self.start = time.time()
+        self.free_space = 0
+        self.remaining = self.total - self.done
+        self.addition = self.done - self.supply
+        self.fraction = min(self.done / self.total, 1.0) if self.total else 0
+        self.elapsed = Timedelta(time.time() - self.start)
+        self.avg = Float(self.elapsed / self.addition if self.addition else 0)
+        self.tta = Timedelta(self.remaining * self.avg)
+
+        line = self.format(self.template)
+
+        self.free_space = (self.columns - len(line) + len(self.animation)
+                           + self.invisible_chars)
+        sys.stdout.write(self.format(line))
+
+        if self.fraction >= 1.0:
+            self.finish()
+        else:
+            sys.stdout.flush()
+
+    def finish(self):
+        sys.stdout.write(self.format(self.outro))
+
+    def __call__(self, **kwargs):
+        self.update(**kwargs)
+
+    def update(self, step=1, **kwargs):
+        if step:
+            self.done += step
+        # Allow to override any properties.
+        self.__dict__.update(kwargs)
+        if self.start is None and 'done' in kwargs:
+            # First call to update and forcing a done value. May be
+            # resuming a download. Keep track for better ETA computation.
+            self.supply = self.done
+        self.render()
+
+    def __next__(self):
+        self.update()
+
+    def iter(self, iterable):
+        for i in iterable:
+            yield i
+            self.update()
+        if self.fraction != 1.0:
+            # Spinner without total.
+            self.finish()
+
+
+# Manage sane default formats while keeping the original type to allow any
+# built-in formatting syntax.
+
+class Float(float):
+
+    def __format__(self, format_spec):
+        if not format_spec:
+            format_spec = '.2f'
+        return super().__format__(format_spec)
+
+
+class Percent(float):
+
+    def __format__(self, format_spec):
+        if not format_spec:
+            format_spec = '.2%'
+        return super().__format__(format_spec)
+
+
+class ETA(datetime):
+
+    def __format__(self, format_spec):
+        if not format_spec:
+            now = datetime.now()
+            diff = self - now
+            format_spec = '%H:%M:%S'
+            if diff.days > 1:
+                format_spec = '%Y-%m-%d %H:%M:%S'
+        return super().__format__(format_spec)
+
+    @classmethod
+    def from_datetime(cls, dt):
+        # Find a more elegant way.
```

Comment:
Agreed, it's weird to have a `classmethod` that initializes the class we inherit from. Maybe just use the `__init__`?